client/lib/signup/step-actions.js

```diff
@@ -346,17 +309,110 @@ export function getUsernameSuggestion( username, reduxState ) {
+export function addPlanToCart( callback, dependencies, stepProvidedItems, reduxStore ) {
+	const { siteSlug } = dependencies;
+	const { cartItem } = stepProvidedItems;
 	if ( isEmpty( cartItem ) ) {
 		// the user selected the free plan
 		defer( callback );
 
 		return;
 	}
 
+	const providedDependencies = { cartItem };
+
 	const newCartItems = [ cartItem ].filter( item => item );
 
-	SignupCart.addToCart( siteSlug, newCartItems, error => callback( error, { cartItem } ) );
+	processItemCart( providedDependencies, newCartItems, callback, reduxStore, siteSlug, null, null );
+}
+
+export function addDomainToCart( callback, dependencies, stepProvidedItems, reduxStore ) {
+	const { siteSlug } = dependencies;
+	const { domainItem, googleAppsCartItem } = stepProvidedItems;
+	const providedDependencies = {
+		siteSlug,
+		domainItem,
+	};
+
+	const newCartItems = [ domainItem, googleAppsCartItem ].filter( item => item );
+
+	processItemCart( providedDependencies, newCartItems, callback, reduxStore, siteSlug, null, null );
+}
+
+function processItemCart(
+	providedDependencies,
+	newCartItems,
+	callback,
+	reduxStore,
+	siteSlug,
+	isFreeThemePreselected,
+	themeSlugWithRepo
+) {
+	const addToCartAndProceed = () => {
+		let privacyItem = null;
+		const state = reduxStore.getState();
+		const { domainItem } = newCartItems;
+
+		if ( domainItem ) {
+			const { product_slug: productSlug } = domainItem;
+			const productsList = getProductsList( state );
+			if ( supportsPrivacyProtectionPurchase( productSlug, productsList ) ) {
+				if ( isDomainTransfer( domainItem ) ) {
+					privacyItem = cartItems.domainTransferPrivacy( {
+						domain: domainItem.meta,
+						source: 'signup',
+					} );
+				} else {
+					privacyItem = cartItems.domainPrivacyProtection( {
+						domain: domainItem.meta,
+						source: 'signup',
+					} );
+				}
+
+				if ( privacyItem ) {
+					newCartItems.push( privacyItem );
+				}
+			}
+		}
+
+		if ( newCartItems.length ) {
+			SignupCart.addToCart( siteSlug, newCartItems, function( cartError ) {
+				callback( cartError, providedDependencies );
+			} );
+		} else {
+			callback( undefined, providedDependencies );
+		}
+	};
+
+	if ( ! user.get() && isFreeThemePreselected ) {
+		setThemeOnSite( addToCartAndProceed, { siteSlug, themeSlugWithRepo } );
+	} else if ( user.get() && isFreeThemePreselected ) {
+		fetchSitesAndUser(
+			siteSlug,
+			setThemeOnSite.bind( null, addToCartAndProceed, { siteSlug, themeSlugWithRepo } ),
+			reduxStore
+		);
+	} else if ( user.get() ) {
+		fetchSitesAndUser( siteSlug, addToCartAndProceed, reduxStore );
+	} else {
+		addToCartAndProceed();
+	}
+}
+
+export function launchSiteApi( callback, dependencies, {}, reduxStore ) {
+	const { siteSlug } = dependencies;
+	const siteId = getSiteId( reduxStore.getState(), siteSlug );
+	const providedDependencies = { siteSlug };
+
+	wpcom.undocumented().launchSite( siteId, function( error ) {
```

Comment:
I think the API will also accept a siteSlug, so we can simplify this code.

Reply:
Implemented in [b2fcae0](https://github.com/Automattic/wp-calypso/pull/30242/commits/b2fcae034db8926a6854201af72a1966f6b32963).